src/core/gen/gen.ml

```diff
@@ -0,0 +1,67 @@
+(* trivial cppo replacement with just enough features to do conditional compilation *)
+
+let if_ocaml_version = Re.(seq [
+    bol; rep blank; str "#if"; rep1 space; str "OCAML_VERSION"; rep1 space; str ">="; rep1 space;
+    char '(';
+    group (rep1 digit); rep1 (set ", "); group (rep1 digit); rep1 (set ", "); group (rep1 digit);
+    char ')'
+  ])
+  |> Re.compile
+
+let version_extract = Re.(seq [
+  group (rep1 digit); char '.'; group (rep1 digit); char '.'; group (rep1 digit)
+  ])
+  |> Re.compile
+
+let current_version =
+  Scanf.sscanf Sys.ocaml_version "%u.%u.%u" (fun a b c -> (a, b, c))
+
+let greater_or_equal (major', minor', patch') =
+  let major, minor, patch = current_version in
+  match major > major', major = major with
+  | true, _ -> true
+  | false, false -> false
+  | false, true -> (
+      match minor > minor', minor = minor' with
+      | true, _ -> true
+      | false, false -> false
+      | false, true -> (
+        match patch > patch', patch = patch' with
+        | true, _ -> true
+        | false, false -> false
+        | false, true -> true))
```

Comment:
You can simplify this using polymorphic comparison on tuples (safe as long as you set the type somewhere):
```suggestion
let greater_or_equal v = (v : (int * int * int)) >= current_version
```